src/csharp/Grpc.Core/Internal/NativeCallbackDispatcher.cs

```diff
@@ -0,0 +1,121 @@
+#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Collections.Generic;
+using Grpc.Core.Logging;
+
+namespace Grpc.Core.Internal
+{
+    internal delegate int UniversalNativeCallback(IntPtr arg0, IntPtr arg1, IntPtr arg2, IntPtr arg3, IntPtr arg4, IntPtr arg5);
+
+    internal delegate int NativeCallbackDispatcherCallback(IntPtr tag, IntPtr arg0, IntPtr arg1, IntPtr arg2, IntPtr arg3, IntPtr arg4, IntPtr arg5);
+
+    internal class NativeCallbackDispatcher
+    {
+        static readonly ILogger Logger = GrpcEnvironment.Logger.ForType<NativeCallbackDispatcher>();
+        static readonly object staticLock = new object();
+        static readonly AtomicCounter atomicCounter = new AtomicCounter();
+        static readonly ConcurrentDictionary<IntPtr, UniversalNativeCallback> registry = new ConcurrentDictionary<IntPtr, UniversalNativeCallback>();
+
+        static NativeCallbackDispatcherCallback dispatcherCallback;
+
+        public static void Init(NativeMethods native)
+        {
+            lock (staticLock)
+            {
+                if (dispatcherCallback == null)
+                {
+                    dispatcherCallback = new NativeCallbackDispatcherCallback(HandleDispatcherCallback);
+                    native.grpcsharp_native_callback_dispatcher_init(dispatcherCallback);
+                }
+            }
+        }
+
+        public static NativeCallbackRegistration RegisterCallback(UniversalNativeCallback callback)
+        {
+            while (true)
+            {
+                // TODO: retries might not work well on 32-bit
```

Comment:
I still need to address this one, in theory finding the next free available tag on 32-bit can take long time. (if more than 2 billion tags are used during the process' lifetime, which is unlikely, but could happen).
On 64-bit we are basically guaranteed never to run out of tags.

Reply:
Very unlikely but do we have a guarantee that the native object allocator is limited to creating `sizeof(long)` objects (the range of `AtomicCounter`). That is, is the C# type `long` always guaranteed to be equal to e.g. the size of C type `void*`?

Reply:
If we can continue to use `GCHandle`  for protecting managed objects from garbage collection, IMO it would be easiest to reason about.

What prevents us from the following:

1) `RegisterCallback(UniversalNativeCallback)` just calls `GCHandle.Alloc` on the callback, and returns that result with something like `GCHandle.ToIntPtr(handle)` - then pass that result as the tag to the native code.

2) `UnregisterCallback` takes back one of those same `IntPtr`, and does something like `GCHandle.FromIntPtr(<param).Free();`

3) Remove the concurrent dictionary mapping tags to `UniversalNativeCallback`

4) In `HandleDispatcherCallback`, do something like: `((UniversalNativeCallback )GCHandle.FromIntPtr(<tag param>).Target)(args....);`

?

Reply:
That sounds great, I updated to PR to use your approach. Thanks for the improvement!
PTAL.